plugins/scaffolder-backend/src/service/router.test.ts

```diff
@@ -102,1634 +82,1224 @@ function createDatabase(): DatabaseService {
+        properties: {
+          requiredParameter1: {
+            type: 'string',
+            description: 'Required parameter 1',
+          },
+        },
+      },
+      {
+        type: 'object',
+        required: ['requiredParameter2'],
+        'backstage:permissions': {
+          tags: ['parameters-tag'],
+        },
+        properties: {
+          requiredParameter2: {
+            type: 'string',
+            description: 'Required parameter 2',
+          },
+        },
+      },
     ],
   },
-  {
-    desc: 'legacy template globals',
-    additionalTemplateGlobals: {
-      nul: null,
-      nop: x => x,
-    } as Record<string, TemplateGlobal>,
+});
+
+const mockUser: UserEntity = {
+  apiVersion: 'backstage.io/v1alpha1',
+  kind: 'User',
+  metadata: {
+    name: 'guest',
+    annotations: {
+      'google.com/email': 'bobby@tables.com',
+    },
   },
-  {
-    desc: 'created template globals',
-    additionalTemplateGlobals: [
-      createTemplateGlobalValue({
-        id: 'nul',
-        description: 'null value',
-        value: null,
-      }),
-      createTemplateGlobalFunction({
-        id: 'nop',
-        description: 'nop function',
-        schema: z =>
-          z
-            .function()
-            .args(z.any().describe('input'))
-            .returns(z.any().describe('output')),
-        fn: x => x,
+  spec: {
+    profile: {
+      displayName: 'Robert Tables of the North',
+    },
+  },
+};
+
+const createTestRouter = async (
+  overrides: {
+    additionalTemplateFilters?:
+      | Record<string, TemplateFilter>
+      | CreatedTemplateFilter<any, any>[];
+    additionalTemplateGlobals?:
+      | Record<string, TemplateGlobal>
+      | CreatedTemplateGlobal[];
+    autocompleteHandlers?: Record<string, AutocompleteHandler>;
+  } = {},
+) => {
+  const logger = mockServices.logger.mock({
+    child: () => logger,
+  });
+
+  const databaseTaskStore = await DatabaseTaskStore.create({
+    database: createDatabase(),
+  });
+
+  const taskBroker = new StorageTaskBroker(databaseTaskStore, logger, config);
+
+  jest.spyOn(taskBroker, 'dispatch');
+  jest.spyOn(taskBroker, 'claim');
+  jest.spyOn(taskBroker, 'cancel');
+  jest.spyOn(taskBroker, 'retry');
+  jest.spyOn(taskBroker, 'list');
+  jest.spyOn(taskBroker, 'get');
+  jest.spyOn(taskBroker, 'vacuumTasks');
+  jest.spyOn(taskBroker, 'event$');
+
+  const catalog = catalogServiceMock.mock();
+  const permissions = mockServices.permissions.mock();
+
+  permissions.authorizeConditional.mockImplementation(async p =>
+    p.map(innerP => ({
+      ...innerP,
+      result: AuthorizeResult.ALLOW,
+    })),
+  );
+
+  permissions.authorize.mockImplementation(async p =>
+    p.map(innerP => ({
+      ...innerP,
+      result: AuthorizeResult.ALLOW,
+    })),
+  );
+
+  const auth = mockServices.auth();
+  const httpAuth = mockServices.httpAuth();
+  const events = mockServices.events();
+
+  catalog.getEntityByRef.mockImplementation(async ref => {
```

Comment:
it's a bit unfortunate that the functional `mockServices.catalog({ entities ... })` doesn't let you make assertions on the methods - I've been meaning to address that so you don't have to switch to the pure mock immediately once you have to check call patterns

Reply:
But in the meantime you may want to use the real mock catalog implementation and using spyOn to grab calls

Reply:
Yeah the default in memory mock should have spies around the methods so you can do both :pray: